radix-engine/src/engine/component_objects.rs

```diff
@@ -0,0 +1,154 @@
+        }
+        for (vid, vault) in unclaimed_lazy_map.descendent_vaults {
+            self.descendent_vaults.insert(vid, vault);
+        }
+    }
+
+    pub fn insert_descendents(&mut self, new_descendents: ComponentObjectsSet) {
+        for (vid, vault) in new_descendents.vaults {
+            self.descendent_vaults.insert(vid, vault);
+        }
+
+        for (mid, child_lazy_map) in new_descendents.lazy_maps {
+            self.insert_map_descendent(child_lazy_map, mid);
+        }
+    }
+}
+
+pub struct ComponentObjectsSetRef {
+    pub mids: HashSet<Mid>,
+    pub vids: HashSet<Vid>,
+}
+
+impl ComponentObjectsSetRef {
+    pub fn new() -> Self {
+        ComponentObjectsSetRef {
+            mids: HashSet::new(),
+            vids: HashSet::new(),
+        }
+    }
+
+    pub fn extend(&mut self, other: ComponentObjectsSetRef) {
+        self.mids.extend(other.mids);
+        self.vids.extend(other.vids);
+    }
+
+    pub fn remove(&mut self, other: &ComponentObjectsSetRef) -> Result<(), RuntimeError> {
+        // Only allow vaults to be added, never removed
+        for vid in &other.vids {
+            if !self.vids.remove(&vid) {
+                return Err(RuntimeError::VaultRemoved(vid.clone()));
+            }
+        }
+
+        for mid in &other.mids {
+            if !self.mids.remove(&mid) {
+                return Err(RuntimeError::LazyMapRemoved(mid.clone()));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Component type objects which will eventually move into a component
+pub struct ComponentObjectsSet {
+    /// Lazy maps which haven't been assigned to a component or lazy map yet.
+    /// Keeps track of vault and lazy map descendents.
+    pub lazy_maps: HashMap<Mid, UnclaimedLazyMap>,
+    /// Vaults which haven't been assigned to a component or lazy map yet.
+    pub vaults: HashMap<Vid, Vault>,
+}
+
+impl ComponentObjectsSet {
+    pub fn new() -> Self {
+        ComponentObjectsSet {
+            lazy_maps: HashMap::new(),
+            vaults: HashMap::new(),
+        }
+    }
+
+    pub fn take(
+        &mut self,
+        other: ComponentObjectsSetRef,
+    ) -> Result<ComponentObjectsSet, RuntimeError> {
+        let mut vaults = HashMap::new();
+        let mut lazy_maps = HashMap::new();
+
+        for vid in other.vids {
+            let vault = self
+                .vaults
+                .remove(&vid)
+                .ok_or(RuntimeError::VaultNotFound(vid))?;
+            vaults.insert(vid, vault);
+        }
+
+        for mid in other.mids {
+            let lazy_map = self
+                .lazy_maps
+                .remove(&mid)
+                .ok_or(RuntimeError::LazyMapNotFound(mid))?;
+            lazy_maps.insert(mid, lazy_map);
+        }
+
+        Ok(ComponentObjectsSet { vaults, lazy_maps })
+    }
+
+    pub fn insert_objects_into_map(&mut self, new_objects: ComponentObjectsSet, mid: &Mid) {
+        let unclaimed_map = self.lazy_maps.get_mut(mid).unwrap();
```

Comment:
RE crashes here when there is a cyclic reference graph in user space, as the root is taken from the process.

Test code:
```rust
use scrypto::prelude::*;
use scrypto::engine::*;

blueprint! {
    struct Hello {
        maps: LazyMap<u32, LazyMap<u32, u32>>
    }

    impl Hello {
        pub fn new() -> Component {
            let map0 = LazyMap::new();
            let map1 = LazyMap::new();
            map0.insert(1u32, map1);

            let input = PutLazyMapEntryInput {
                mid: Mid(Context::transaction_hash(), 1025),
                key: scrypto_encode(&0u32),
                value: scrypto_encode(&Mid(Context::transaction_hash(), 1024)),
            };
            let _: PutLazyMapEntryOutput = call_engine(PUT_LAZY_MAP_ENTRY, input);
            
            Hello {
                maps: map0
            }.instantiate()
        }
    }
}
```